File: code/scenes/src/main/java/org/fbme/scenes/controllers/ROLayoutModelBase.java
package org.fbme.scenes.controllers;

import org.jetbrains.annotations.NotNull;

import java.util.HashSet;
import java.util.Set;

public abstract class ROLayoutModelBase<T> implements ROLayoutModel<T> {

    private final Set<ROLayoutModel.Listener<T>> myListeners = new HashSet<ROLayoutModel.Listener<T>>();

    protected final void fireMoved(@NotNull Set<T> components, int dx, int dy, boolean completed) {
        for (ROLayoutModel.Listener<T> listener : myListeners) {
            listener.onComponentsMoved(components, dx, dy, completed);
        }
    }

    protected final void fireAdded(@NotNull T component, int x, int y) {
        for (ROLayoutModel.Listener<T> listener : myListeners) {
            listener.onComponentAdded(component, x, y);
        }
    }

    protected final void fireRemoved(@NotNull T component) {
        for (ROLayoutModel.Listener<T> listener : myListeners) {
            listener.onComponentRemoved(component);
        }
    }

    @Override
    public final void addListener(@NotNull ROLayoutModel.Listener<T> listener) {
        myListeners.add(listener);
    }

    @Override
    public final void removeListener(@NotNull ROLayoutModel.Listener<T> listener) {
        myListeners.remove(listener);
    }
}

File: code/scenes/src/main/kotlin/org/fbme/scenes/controllers/ROLayoutModelBase.kt
package org.fbme.scenes.controllers

abstract class ROLayoutModelBase<T> : ROLayoutModel<T> {
    private val listeners: MutableSet<ROLayoutModel.Listener<T>> = HashSet()
    protected fun fireMoved(components: Set<T>, dx: Int, dy: Int, completed: Boolean) {
        for (listener in listeners) {
            listener.onComponentsMoved(components, dx, dy, completed)
        }
    }

    protected fun fireAdded(component: T, x: Int, y: Int) {
        for (listener in listeners) {
            listener.onComponentAdded(component, x, y)
        }
    }

    protected fun fireRemoved(component: T) {
        for (listener in listeners) {
            listener.onComponentRemoved(component)
        }
    }

    override fun addListener(listener: ROLayoutModel.Listener<T>) {
        listeners.add(listener)
    }

    override fun removeListener(listener: ROLayoutModel.Listener<T>) {
        listeners.remove(listener)
    }
}
